File: TinyScript/java/src/main/java/parser/ast/DeclareStmt.java
package parser.ast;

import parser.util.ParseException;
import parser.util.PeekTokenIterator;

public class DeclareStmt extends Stmt {
    public DeclareStmt() {
        super(ASTNodeTypes.DECLARE_STMT, "declare");
    }

    public static ASTNode parse(PeekTokenIterator it) throws ParseException {
        var stmt = new DeclareStmt();
        it.nextMatch("var");
        var tkn = it.peek();
        var factor = Factor.parse(it);
        if(factor == null) {
           throw new ParseException(tkn);
        }
        stmt.addChild(factor);
        var lexeme = it.nextMatch("=");
        var expr = Expr.parse(it);
        stmt.addChild(expr);
        stmt.setLexeme(lexeme);
        return stmt;

    }
}

File: TinyScript/js/src/parser/ast/DeclareStmt.js
const ASTNodeTypes = require('./ASTNodeTypes')
const {Stmt} = require('./index')
class DeclareStmt extends Stmt {
    constructor() {
        super(ASTNodeTypes.DECLARE_STMT, 'declare')
    }

}

module.exports = DeclareStmt 
const {Factor, Expr} = require('./index')

DeclareStmt.parse = (it) => {
    const stmt = new DeclareStmt()
    it.nextMatch("var")
    const tkn = it.peek()
    const factor = Factor.parse(it)
    if(factor == null) {
        throw ParseException.fromToken(tkn)
    }
    stmt.addChild(factor)
    const lexeme = it.nextMatch("=")
    const expr = Expr.parse(it)
    stmt.addChild(expr)
    stmt.setLexeme(lexeme)
    return stmt
}